### PR TITLE
IDEA-157915: Fix symbolic-ref in HEAD

### PR DESCRIPTION
**Pull request by illabefat**

### CODE REVIEW

**Comment by klikh** [comment]
Have you submitted a contributor license agreement? If not, please check your mail box: I've just sent a sign request there via DocuSign.

**Comment by illabefat** [comment]
Yes, I've just signed it.

**Comment by klikh** [comment]
Received, thanks. 

I see that you've at first closed this pull request right after creation, and then reopened it a week later. But Github doesn't show any new/rewritten commits here. You just wanted to test it more, or you closed it by accident?

**Comment by illabefat** [comment]
Yes, I just wanted PR to be first reviewed by my colleagues, but accidentally created it in JetBrains repo instead of my fork

File: plugins/git4idea/src/git4idea/repo/GitRepositoryFiles.java [hunk 1]
```diff
@@ -316,4 +316,9 @@ public void refresh(boolean async) {
   Collection<VirtualFile> getRootDirs() {
     return ContainerUtil.newHashSet(myMainDir, myWorktreeDir);
   }
+
+  @NotNull
+  File getBranchFile(String path) {
+    return new File(FileUtil.toSystemDependentName(myMainDir.getCanonicalPath() + File.separatorChar + path));
```
Comment (klikh) [inline plugins/git4idea/src/git4idea/repo/GitRepositoryFiles.java]: Looks like this should be changed to 
```
  @NotNull
  File getBranchFile(@NotNull String fullBranchName) {
    return file(myMainDir.getPath() + slash(fullBranchName));
  }
```
(see other `get*File` methods and `isBranchFile()`)

File: plugins/git4idea/src/git4idea/repo/GitRepositoryReader.java [hunk 2]
```diff
@@ -331,11 +337,22 @@ private HeadInfo readHead() {
 
     Hash hash = parseHash(headContent);
     if (hash != null) {
-      return new HeadInfo(false, headContent);
+      if (alreadyVisited.isEmpty()) {
+        return new HeadInfo(false, headContent);
+      } else {
+        return new HeadInfo(true, alreadyVisited.get(alreadyVisited.size() - 1));
```
Comment (klikh) [inline plugins/git4idea/src/git4idea/repo/GitRepositoryReader.java]: We've got a handy `ContainerUtil.getLastItem()` for this.

**Comment by klikh** [comment]
Merged the patch with a few mentioned corrections. Thanks a lot!

**Comment by klikh** [comment]
Ah. Together we missed the fact that the linked reference can reside in `.git/packed-refs`. Currently it leads to `RepoStateException: Couldn't load file /Users/loki/temp/t/.git/refs/heads/master`.

**Comment by illabefat** [comment]
And how can this happen? symbolic-ref in origin?

**Comment by klikh** [comment]
`git pack-refs --all` makes everything packed, including `refs/heads/master` thus the code can't find `.git/refs/heads/master` written in `.git/HEAD`. 

It seems that to solve the issue we'll have first to read all references to a map, both from `.git/refs` and `.git/packed-refs`, and then resolve them by following links inside the map.

**Comment by illabefat** [comment]
Thank you, I'll try to do something with this.

**Comment by klikh** [comment]
Cool, thanks. I've written a test which reveals the issue. 

```
package git4idea.repo

import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.vfs.LocalFileSystem
import git4idea.branch.GitBranchUtil.stripRefsPrefix
import git4idea.test.GitExecutor.cd
import git4idea.test.GitExecutor.git
import git4idea.test.GitPlatformTest
import git4idea.test.GitTestUtil.makeCommit
import java.io.File

class GitRepositoryReaderNew2Test : GitPlatformTest() {

  fun `test packed symbolic ref`() {
    cd(myProjectPath)
    git("init")
    makeCommit("file.txt")
    git("pack-refs --all")

    // verify master is packed
    val gitDir = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(File(myProjectPath, ".git"))!!
    val repoFiles = GitRepositoryFiles.getInstance(gitDir)

    assertFalse("setup failed: master is not packed", File(repoFiles.refsHeadsFile, "master").exists())
    assertTrue("setup failed: master is not packed",
               FileUtil.loadFile(repoFiles.packedRefsPath).lines().any { it.endsWith("refs/heads/master") })

    val repo = GitRepositoryImpl.getInstance(myProjectRoot, myProject, false)
    assertEquals("HEAD should be on master", "master", stripRefsPrefix(repo.currentBranchName!!))
    assertEquals("HEAD revision is incorrect", git("rev-parse HEAD"), repo.currentRevision)
    assertEquals("master revision is incorrect",
                 git("rev-parse master"),
                 repo.branches.getHash(repo.branches.findLocalBranch("master")!!)!!.asString())
  }
}
```

**Pull request closed without merging**